### ISSUE TITLE
IPVS is non-functional in 1.26

### ISSUE THREAD

**Issue by liaralabs** [k3s-io/k3s#7183]
**Environmental Info:**
K3s Version: 
k3s version v1.26.3+k3s1 (01ea3ff2)
go version go1.19.7

Node(s) CPU architecture, OS, and Version: 
Debian 11, amd64, hetzner cloud VPS
Linux debian-2gb-hil-1 5.10.0-21-amd64 #1 SMP Debian 5.10.162-1 (2023-01-21) x86_64 GNU/Linux

Cluster Configuration:
Single node

**Describe the bug:**
Starting with 1.26 having IPVS enabled completely breaks the cluster's traffic flow. As soon as IPVS is enabled and the firewall rules/ipvsadm policies kick in, the internal cluster communication breaks in addition to external traffic to rather important ports such as port 22 and port 6443

No pods are able to spin up, attempting to install calico results in the pod not being able to communicate with the API server and the cluster is in a largely failed state.

**Steps To Reproduce:**

Spin up a fresh Debian 11 VPS -- I used hetzner for this test

Install requisites for ipvs:

```
apt install ipset ipvsadm
```

Install k3s and enable ipvs all at once
```
curl -sfL https://get.k3s.io | sh -s - server --kube-proxy-arg proxy-mode=ipvs
```

Wait for initialization -- as soon as iptables ipvs policies are enacted, the server is now broken an unable to accept new ingress connections.

**Expected behavior:**
The cluster works.

I can ssh into my machine with ipvs enabled

I can use the kube api externally with ipvs enabled

I can start pods (i.e. Calico) and finish building up my cluster (note, flannel was left enabled for this test to prove how simple the error is to reproduce)

**Actual behavior:**
No new connections can be created.

Cannot create a new SSH connection.

Cannot connect to the kube api server from an external connection

Cannot ping the external IP from the command line of the server itself or an external IP

Cannot spin up new pods or install calico

**Additional context / logs:**

I do not have such issues with IPVS in my vanilla k8s cluster running 1.26.1

The issue was not present in 1.25. Downgrading a broken machine to the latest 1.25 branch results in the behavior disappearing.

It appears the difference between 1.25 and 1.26 is coming down to iptables rules

These iptables rules appear new as of 1.26 (they are not present in k3s 1.25.8):

```
-A KUBE-IPVS-FILTER -m set --match-set KUBE-LOAD-BALANCER dst,dst -j RETURN
-A KUBE-IPVS-FILTER -m set --match-set KUBE-CLUSTER-IP dst,dst -j RETURN
-A KUBE-IPVS-FILTER -m set --match-set KUBE-EXTERNAL-IP dst,dst -j RETURN
-A KUBE-IPVS-FILTER -m conntrack --ctstate NEW -m set --match-set KUBE-IPVS-IPS dst -j REJECT --reject-with icmp-port-unreachable
```

Removing the rule allows remote connections to once again be initiated:
```
-A KUBE-IPVS-FILTER -m conntrack --ctstate NEW -m set --match-set KUBE-IPVS-IPS dst -j REJECT --reject-with icmp-port-unreachable
```

My vanilla 1.26.1 cluster has the same rules and has none of these issues.

Finally, the issue appears to have silently snuck in during my node's automatic update to 1.26 (my mistake). It's curious to note that nginx ingress traffic on port 80 and 443 were unaffected while the server was still running (though ssh/kube api was already dead) -- that is until I rebooted my server and calico was unable to start, which cascaded into a full failure because no services were able to start upon a reboot.

**Comment by brandond** [comment]
> No pods are able to spin up, attempting to install calico results in the pod not being able to communicate with the API server and the cluster is in a largely failed state.
> I can start pods (i.e. Calico) and finish building up my cluster (note, flannel was left enabled for this test to prove how simple the error is to reproduce)

I'm confused, are you using both the embedded flannel **and** calico at the same time?

**Comment by liaralabs** [comment]
> I'm confused, are you using both the embedded flannel and calico at the same time?

No I have two different clusters. The single node cluster I use for production tasks broke following the update. This instance uses calico.

The POC for reproduction does not use flannel and was kept as simple as possible to remove possibility of conflicting issues as a result of my production instance's more complex configuration. Apologies if that was unclear

> These aren't rules we add ourselves, these rules would be managed by kube-proxy. Do you perhaps need to add additional kube-proxy args to accomplish what you're trying to do?

I have wondered this myself; however, all k3s documentation I can find related to ipvs should note that simply enabling the option in kube-proxy is enough

**Comment by brandond** [comment]
Have you compared the args passed to kube-proxy between k3s and your vanilla (by which I'm assuming you mean kubeadm) cluster? That is probably where I'd start. None of the kube-proxy code itself is modified by this project.

**Comment by liaralabs** [comment]
Thanks for the hint in the right direction. Upon closer examination, I was not able to determine any specific configuration settings of kube-proxy responsible for this; however, I did note that the issue seems to come down to the related ipset for KUBE-IPVS-IPS (i.e. the IPSET matched against for rejections).

The ipset for the kubeadm cluster, which has no outward issues:

```
Name: KUBE-IPVS-IPS
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 1592
References: 1
Number of entries: 29
Members:
10.96.39.198
10.96.225.74
10.96.241.27
10.96.46.218
10.96.0.1
10.96.111.91
10.96.164.151
10.96.214.183
10.96.71.208
10.96.147.63
10.96.51.94
10.96.157.157
10.96.224.160
10.96.148.141
10.96.40.187
10.96.3.60
10.96.140.173
10.96.186.192
10.96.113.19
10.96.2.160
10.96.239.154
10.96.61.139
10.96.213.159
10.96.0.10
10.96.197.245
10.96.40.159
10.96.174.88
10.96.57.126
10.96.11.132
```

The ipset of the Debian 11 POC k3s instance:

```
Name: KUBE-IPVS-IPS
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 440
References: 1
Number of entries: 5
Members:
10.43.0.1
10.43.0.10
10.43.238.164
10.43.36.168
5.xx.xx.xxx <-- external IP
```

It seems to me the major config difference is that some process (unknown to me) has determined the external IP should be passed to the KUBE-IPVS-IPS ipset, which gets matched against rejection.

I'm only able to influence the address passed to this ipset by altering the `advertise-address` of the server (subsequently `node-ip` and `external-node-ip` influence this value by being the value advertise-address defaults to. Whatever value ends up here ends up being passed to the IPVS ipset list  and gets blocked by the iptables rule first mentioned.

That said, the external IP of a single-node cluster is the most logical advertise address since no private IPs exist by default in such a setup -- other than 127.0.0.1 (for which pinging localhost will then be blocked if we change the advertise address to localhost)

This directly contrasts the behavior of a kubeadm k8s cluster, whereby the advertiseAddress configured to the cluster is not passed to IPVS

```
apiVersion: kubeadm.k8s.io/v1beta3
kind: InitConfiguration  
localAPIEndpoint:  
  advertiseAddress: "10.0.0.2"  
nodeRegistration:
  criSocket: unix:///var/run/crio/crio.sock
  kubeletExtraArgs:  
    node-ip: "10.0.0.2"
    cloud-provider: "external"
---
apiVersion: kubeadm.k8s.io/v1beta3
kind: ClusterConfiguration
apiServer:
  certSANs:
  - "5.external.ip"
  - "10.0.0.2"
networking:
  podSubnet: "10.85.0.0/16"
  serviceSubnet: "10.96.0.0/16"
```

And all the ipvs managed IPs -- no 10.0.0.2 here
```
kube-ipvs0       DOWN           10.96.157.157/32 10.96.0.10/32 10.96.11.132/32 10.96.186.192/32 10.96.40.159/32 10.96.197.245/32 10.96.147.63/32 10.96.214.183/32 10.96.213.159/32 10.96.164.151/32 10.96.111.91/32 10.96.39.198/32 10.96.71.208/32 10.96.51.94/32 10.96.57.126/32 10.96.0.1/32 10.96.225.74/32 10.96.140.173/32 10.96.113.19/32 10.96.241.27/32 10.96.61.139/32 10.96.2.160/32 10.96.239.154/32 10.96.40.187/32 10.96.174.88/32 10.96.148.141/32 10.96.3.60/32 10.96.224.160/32 10.96.46.21
```

It may not be an issue with kube-proxy directly, but some process in the spin up of the cluster thinks the external IP should be treated as an internal pod address and ends up passing this IP to IPVS as the ipsets are being created/sync'd

**Comment by liaralabs** [comment]
Some further musings (apologies, I keep digging deeper as my understanding of the issue grows and some new things get uncovered).

ipvs adding the node's external IP to the IPVS interface appears a result of the traefik ingress mixed with the servicelb load balancer. Kubernetes is supposed to only listen to services to add to IPVS, which clued me in to go looking for services bound to external IPs.

In the POC instance, that's the traefik load balancer, in my previously working production instance, that's a metallb load balancer for nginx ingress. Curiously, the kubeadm cluster's loadbalancer does not display this behavior. Its loadbalancer is a physical instance, but does have an external IP assigned to it. Curiously the internal/external IP of this load balancer are not discovered by IPVS, despite a load balancer service present, pointing to a DNS entry which resolves to the external IP. So, I'm still confused about this discrepancy between the clusters but am glad to at least have a better handle of the underlying mechanisms at play.

I will have to keep digging though I'm no longer certain this is specific to k3s. I would have to setup a single node kubeadm cluster/metallb to test this, but that is not a task for today.

**Comment by brandond** [comment]
Do you by any chance have a service on that address, one using ServiceLB or otherwise? Kube-proxy handles routing traffic to service ClusterIP addresses, I wonder if for some reason it is also adding the service's loadbalancer address to the ipset. ServceLB is the primary thing that K3s would have but your kubeadm cluster would not.

**Comment by liaralabs** [comment]
> Kube-proxy handles routing traffic to service ClusterIP addresses, I wonder if for some reason it is also adding the service's loadbalancer address to the ipset. ServiceLB is the primary thing that K3s would have but your kubeadm cluster would not.

We responded very closely to each other :slightly_smiling_face: That indeed appears to be exactly what's happening. In the POC cluster, the ip assigned to the traefik ingress service via service LB gets discovered by IPVS. This doesn't seem desirable but I'm not familiar enough with the underlying code to know if that is "correct"

**Comment by brandond** [comment]
I wouldn't expect it to pick up the loadbalancer address; I've not seen it do that nor received reports of problems with ipvs in the past so perhaps this is new behavior from kube-proxy. I'd have to do some research to see if it's expected or configurable.

**Comment by brandond** [comment]
@rbrtbnfgl @thomasferrandiz do you have any clues on whether or not this is something that might have changed in recent versions? I will admit I haven't ever played with IPVS much. I will say that I wouldn't expect it to add IPVS entries for external loadbalancer service addresses but its possible this is something that's changed upstream?

**Comment by liaralabs** [comment]
Seems like this was actually an issue with upstream despite me having issues replicating it in my kubeadm built cluster.

The threads linked here are showing that a PR has been merged and the fix is targeted for the 1.28 release (unless it gets backported).

Thanks for the assistance in helping to track this down.

**Comment by liaralabs** [comment]
The original fix was implemented in 1.28 and backported in 1.27.2. A fix hasn't yet made it into 1.26.

This could be closed because it is resolved in newer versions, but it is also still not functional if for some reason you can't upgrade.

**Comment by brandond** [comment]
Closing as resolved upstream; it'd be nice if they would decide to backport it to 1.26.

**Comment by michal0000000** [comment]
Issue is still present in 2.28.3 as far as I can tell.

**Comment by SoulKyu** [comment]
Look like your issue is relate to : https://github.com/kubernetes/kubernetes/pull/115019#issuecomment-1784984211
